Cannot use InjectMock with gRCP generated classes

### Describe the bug

We're trying to mock a gRPC client using `InjectMock` like so:

```java
@InjectMock
DeviceService client;
```

`DeviceService` is one of the autogenerated gRPC classes, and specifically is defined like this  
`public interface DeviceService extends MutinyService`.

Running the test fails with: 

```
org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
Caused by: java.lang.reflect.InvocationTargetException
Caused by: java.lang.IllegalStateException: Invalid use of io.quarkus.test.junit.mockito.InjectMock - the injected bean does not declare a CDI normal scope but: javax.inject.Singleton. Offending field is client of test class class org.project.package.DeviceGatewayTest
```

Since `DeviceService` is a Singleton, then as per the docs https://quarkus.io/guides/getting-started-testing#further-simplification-with-injectmock we should use `convertScopes = true`.
 
 ```java
@InjectMock(convertScopes = true)
DeviceService client;
```
Doing so, produces the same error.

### Expected behavior

`InjectMock(convertScopes = true)` should mock the generated gRPC sources, since these have Singleton scope.

### Actual behavior

Tests fail with 
```
org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
Caused by: java.lang.reflect.InvocationTargetException
Caused by: java.lang.IllegalStateException: Invalid use of io.quarkus.test.junit.mockito.InjectMock - the injected bean does not declare a CDI normal scope but: javax.inject.Singleton. Offending field is client of test class class org.project.package.DeviceGatewayTest
```


### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

`Linux homer 5.13.0-27-generic #29~20.04.1-Ubuntu SMP Fri Jan 14 00:32:30 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

### Output of `java -version`

```
openjdk version "11.0.13" 2021-10-19  
OpenJDK Runtime Environment (build 11.0.13+8-Ubuntu-0ubuntu1.20.04)  
OpenJDK 64-Bit Server VM (build 11.0.13+8-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)
```

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

`2.2.2.Final`
(Also tested with `2.7.1.Final`, still the same.)

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

```
Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)  
Maven home: /home/chris/.m2/wrapper/dists/apache-maven-3.8.1-bin/2l5mhf2pq2clrde7f7qp1rdt5m/apache-maven-3.8.1  
Java version: 11.0.13, vendor: Ubuntu, runtime: /usr/lib/jvm/java-11-openjdk-amd64  
Default locale: en_US, platform encoding: UTF-8  
OS name: "linux", version: "5.13.0-27-generic", arch: "amd64", family: "unix"
```

### Additional information

_No response_

--- a/docs/src/main/asciidoc/grpc-service-consumption.adoc
+++ b/docs/src/main/asciidoc/grpc-service-consumption.adoc
@@ -404,3 +404,28 @@ By default, when starting the application in dev mode, a gRPC server is started,
 You can configure the gRPC extension's dev mode behavior using the following properties.
 
 include::{generated-dir}/config/quarkus-grpc-config-group-grpc-dev-mode-config.adoc[opts=optional, leveloffset=+1]
+
+== Inject mock clients
+
+In your `@QuarkuTest`, you can use `@InjectMock` to inject the Mutiny client of a gRPC service:
+
+[source, java]
+----
+@QuarkusTest
+public class GrpcMockTest {
+
+    @InjectMock
+    @GrpcClient("hello")
+    Greeter greeter;
+
+    @Test
+    void test1() {
+        HelloRequest request = HelloRequest.newBuilder().setName("neo").build();
+        Mockito.when(greeter.sayHello(Mockito.any(HelloRequest.class)))
+                .thenReturn(Uni.createFrom().item(HelloReply.newBuilder().setMessage("hello neo").build()));
+        Assertions.assertEquals(greeter.sayHello(request).await().indefinitely().getMessage(), "hello neo");
+    }
+}
+----
+
+IMPORTANT: Only the Mutiny client can be _mocked_, channels, and other stubs cannot be mocked.

--- a/extensions/grpc/api/pom.xml
+++ b/extensions/grpc/api/pom.xml
@@ -12,6 +12,10 @@
 
     <dependencies>
         <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc</artifactId>
+        </dependency>
+        <dependency>
             <groupId>io.grpc</groupId>
             <artifactId>grpc-stub</artifactId>
             <exclusions>

--- a/extensions/grpc/api/src/main/java/io/quarkus/grpc/GrpcClientUtils.java
+++ b/extensions/grpc/api/src/main/java/io/quarkus/grpc/GrpcClientUtils.java
@@ -3,6 +3,7 @@ package io.quarkus.grpc;
 import io.grpc.Metadata;
 import io.grpc.stub.AbstractStub;
 import io.grpc.stub.MetadataUtils;
+import io.quarkus.arc.ClientProxy;
 
 /**
  * gRPC client utilities
@@ -24,6 +25,9 @@ public class GrpcClientUtils {
         if (client == null) {
             throw new NullPointerException("Cannot attach headers to a null client");
         }
+
+        client = getProxiedObject(client);
+
         if (client instanceof AbstractStub) {
             return (T) MetadataUtils.attachHeaders((AbstractStub) client, extraHeaders);
         } else if (client instanceof MutinyClient) {
@@ -33,5 +37,14 @@ public class GrpcClientUtils {
         } else {
             throw new IllegalArgumentException("Unsupported client type " + client.getClass());
         }
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T> T getProxiedObject(T client) {
+        // If we get a proxy, get the actual instance.
+        if (client instanceof ClientProxy) {
+            client = (T) ((ClientProxy) client).arc_contextualInstance();
+        }
+        return client;
     }
 }

--- a/extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcClientProcessor.java
+++ b/extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcClientProcessor.java
@@ -4,6 +4,7 @@ import static io.quarkus.deployment.Feature.GRPC_CLIENT;
 import static io.quarkus.grpc.deployment.GrpcDotNames.ADD_BLOCKING_CLIENT_INTERCEPTOR;
 import static io.quarkus.grpc.deployment.GrpcDotNames.CONFIGURE_STUB;
 import static io.quarkus.grpc.deployment.GrpcDotNames.CREATE_CHANNEL_METHOD;
+import static io.quarkus.grpc.deployment.GrpcDotNames.MUTINY_CLIENT;
 import static io.quarkus.grpc.deployment.GrpcDotNames.RETRIEVE_CHANNEL_METHOD;
 import static io.quarkus.grpc.deployment.GrpcInterceptors.MICROMETER_INTERCEPTORS;
 import static io.quarkus.grpc.deployment.ResourceRegistrationUtils.registerResourcesForProperties;
@@ -19,6 +20,7 @@ import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.regex.Pattern;
 
+import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.inject.spi.DeploymentException;
 import javax.inject.Singleton;
 
@@ -61,6 +63,7 @@ import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.MethodDescriptor;
 import io.quarkus.gizmo.ResultHandle;
 import io.quarkus.grpc.GrpcClient;
+import io.quarkus.grpc.MutinyClient;
 import io.quarkus.grpc.RegisterClientInterceptor;
 import io.quarkus.grpc.deployment.GrpcClientBuildItem.ClientInfo;
 import io.quarkus.grpc.deployment.GrpcClientBuildItem.ClientType;
@@ -248,9 +251,12 @@ public class GrpcClientProcessor {
                     String clientName = client.getClientName();
                     ExtendedBeanConfigurator configurator = SyntheticBeanBuildItem.configure(clientInfo.className)
                             .addQualifier().annotation(GrpcDotNames.GRPC_CLIENT).addValue("value", clientName).done()
-                            .scope(Singleton.class)
+                            // Only the mutiny client can use the Application scope, the others are "final" and so need Singleton.
+                            // Using @ApplicationScoped allows the usage of @InjectMock
+                            .scope(clientInfo.type == ClientType.MUTINY_CLIENT ? ApplicationScoped.class : Singleton.class)
                             .unremovable()
                             .forceApplicationClass()
+                            .addType(MutinyClient.class)
                             .creator(new Consumer<>() {
                                 @Override
                                 public void accept(MethodCreator mc) {

--- a/extensions/grpc/deployment/src/test/java/io/quarkus/grpc/client/ClientServiceInterfaceCompressionTest.java
+++ b/extensions/grpc/deployment/src/test/java/io/quarkus/grpc/client/ClientServiceInterfaceCompressionTest.java
@@ -14,6 +14,7 @@ import io.grpc.examples.helloworld.Greeter;
 import io.grpc.examples.helloworld.GreeterClient;
 import io.grpc.examples.helloworld.GreeterGrpc;
 import io.quarkus.grpc.GrpcClient;
+import io.quarkus.grpc.GrpcClientUtils;
 import io.quarkus.grpc.server.services.HelloService;
 import io.quarkus.test.QuarkusUnitTest;
 
@@ -31,7 +32,7 @@ public class ClientServiceInterfaceCompressionTest {
 
     @Test
     public void testCallOptions() {
-        GreeterClient client = (GreeterClient) consumer.service;
+        GreeterClient client = (GreeterClient) GrpcClientUtils.getProxiedObject(consumer.service);
         assertEquals("gzip", client.getStub().getCallOptions().getCompressor());
     }
 

--- a/extensions/grpc/deployment/src/test/java/io/quarkus/grpc/client/deadline/ClientDeadlineTest.java
+++ b/extensions/grpc/deployment/src/test/java/io/quarkus/grpc/client/deadline/ClientDeadlineTest.java
@@ -20,6 +20,7 @@ import io.grpc.examples.helloworld.GreeterGrpc;
 import io.grpc.examples.helloworld.HelloReply;
 import io.grpc.examples.helloworld.HelloRequest;
 import io.quarkus.grpc.GrpcClient;
+import io.quarkus.grpc.GrpcClientUtils;
 import io.quarkus.test.QuarkusUnitTest;
 
 public class ClientDeadlineTest {
@@ -36,7 +37,7 @@ public class ClientDeadlineTest {
 
     @Test
     public void testCallOptions() {
-        GreeterClient client = (GreeterClient) consumer.service;
+        GreeterClient client = (GreeterClient) GrpcClientUtils.getProxiedObject(consumer.service);
         Deadline deadline = client.getStub().getCallOptions().getDeadline();
         assertNotNull(deadline);
         HelloReply reply = client.sayHello(HelloRequest.newBuilder().setName("Scaladar").build()).onFailure()

--- a/integration-tests/grpc-plain-text-mutiny/pom.xml
+++ b/integration-tests/grpc-plain-text-mutiny/pom.xml
@@ -48,6 +48,11 @@
     </dependency>
     <dependency>
       <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-junit5-mockito</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
       <artifactId>quarkus-test-grpc</artifactId>
       <version>${project.version}</version> <!--kept here to not pollute the BOM-->
       <scope>test</scope>

--- a/integration-tests/grpc-plain-text-mutiny/src/test/java/io/quarkus/grpc/examples/hello/GrpcMockTest.java
+++ b/integration-tests/grpc-plain-text-mutiny/src/test/java/io/quarkus/grpc/examples/hello/GrpcMockTest.java
@@ -1,0 +1,56 @@
+package io.quarkus.grpc.examples.hello;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+import examples.Greeter;
+import examples.HelloReply;
+import examples.HelloRequest;
+import io.quarkus.grpc.GrpcClient;
+import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.junit.mockito.InjectMock;
+import io.smallrye.mutiny.Uni;
+
+@QuarkusTest
+public class GrpcMockTest {
+
+    @InjectMock
+    @GrpcClient("hello")
+    Greeter greeter;
+
+    @Inject
+    BeanCallingservice beanCallingService;
+
+    @Test
+    void test1() {
+        HelloRequest request = HelloRequest.newBuilder().setName("clement").build();
+        Mockito.when(greeter.sayHello(Mockito.any(HelloRequest.class)))
+                .thenReturn(Uni.createFrom().item(HelloReply.newBuilder().setMessage("hello clement").build()));
+        Assertions.assertEquals(greeter.sayHello(request).await().indefinitely().getMessage(), "hello clement");
+    }
+
+    @Test
+    void test2() {
+        HelloRequest request = HelloRequest.newBuilder().setName("roxanne").build();
+        Mockito.when(greeter.sayHello(request))
+                .thenReturn(Uni.createFrom().item(HelloReply.newBuilder().setMessage("hello roxanne").build()));
+        Assertions.assertEquals(beanCallingService.call(), "hello roxanne");
+    }
+
+    @ApplicationScoped
+    public static class BeanCallingservice {
+        @InjectMock
+        @GrpcClient("hello")
+        Greeter greeter;
+
+        public String call() {
+            return greeter.sayHello(HelloRequest.newBuilder().setName("roxanne").build())
+                    .map(HelloReply::getMessage)
+                    .await().indefinitely();
+        }
+    }
+}

